tools/wasi/wasi.cpp

```diff
@@ -0,0 +1,235 @@
+// Fizzy: A fast WebAssembly interpreter
+// Copyright 2019-2020 The Fizzy Authors.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "execute.hpp"
+#include "parser.hpp"
+#include <uvwasi.h>
+#include <cassert>
+#include <filesystem>
+#include <fstream>
+#include <iostream>
+
+namespace
+{
+// Global state.
+// NOTE: we do not care about uvwasi_destroy(), because it only clears file descriptors (currently)
+// and we are a single-run tool. This may change in the future and should reevaluate.
+uvwasi_t state;
+
+fizzy::execution_result wasi_return_enosys(fizzy::Instance&, std::vector<uint64_t>, int)
+{
+    return {false, {UVWASI_ENOSYS}};
+}
+
+fizzy::execution_result wasi_proc_exit(fizzy::Instance&, std::vector<uint64_t> args, int)
+{
+    uvwasi_proc_exit(&state, static_cast<uvwasi_exitcode_t>(args[0]));
+    // Should not reach this.
+    return {false, {}};
+}
+
+fizzy::execution_result wasi_fd_write(fizzy::Instance& instance, std::vector<uint64_t> args, int)
+{
+    uint32_t fd = static_cast<uint32_t>(args[0]);
+    uint32_t iovptr = static_cast<uint32_t>(args[1]);
+    uint32_t iovcnt = static_cast<uint32_t>(args[2]);
+    uint32_t nwritten = static_cast<uint32_t>(args[3]);
+
+    std::vector<uvwasi_ciovec_t> iovs(iovcnt);
+    // TODO: not sure what to pass as end, passing memory size...
+    uvwasi_errno_t ret = uvwasi_serdes_readv_ciovec_t(instance.memory->data(),
+        instance.memory->size(), iovptr, iovs.data(), static_cast<uvwasi_size_t>(iovs.size()));
```

Comment:
Maybe just using `iovcnt` without cast is simpler
```suggestion
        instance.memory->size(), iovptr, iovs.data(), iovcnt);
```

Reply:
```
/Users/alex/Projects/fizzy/tools/wasi/wasi.cpp:42:60: error: implicit conversion loses integer precision: 'std::__1::vector<uvwasi_ciovec_s, std::__1::allocator<uvwasi_ciovec_s>
      >::size_type' (aka 'unsigned long') to 'uvwasi_size_t' (aka 'unsigned int') [-Werror,-Wshorten-64-to-32]
        instance.memory->size(), iovptr, iovs.data(), iovs.size());
                                                      ~~~~~^~~~~~
```

Reply:
Yes, but `iovcnt` has the right type and equals to this size.

Reply:
Ah didn't see you proposed `iovcnt`.